crates/bevy_ecs/src/component.rs

```diff
@@ -113,6 +115,161 @@ pub trait Component: Send + Sync + 'static {
+
+// SAFETY: There is only one ComponentId, so the order does not matter.
+unsafe impl<T> Bundle for Unlocked<T>
+where
+    T: Component,
+{
+    fn component_ids(
+        components: &mut Components,
+        storages: &mut crate::storage::Storages,
+    ) -> Vec<ComponentId> {
+        vec![components.init_component::<T>(storages)]
+    }
+    unsafe fn from_components<U, F>(ctx: &mut U, mut func: F) -> Self
+    where
+        F: FnMut(&mut U) -> crate::ptr::OwningPtr,
+    {
+        Self(func(ctx).read::<T>())
+    }
+    fn get_components(self, func: impl FnMut(crate::ptr::OwningPtr)) {
+        crate::ptr::OwningPtr::make(self.0, func);
+    }
+}
+
+/// A [`Bundle`] type that can contain a [`Component`] with protected write access.
+///
+/// The resulting `Unlocked<T, Marker>` type is itself private, so it must be encapsulated
+/// somehow, which prevents outside modification of the protected component.
+///
+/// # Examples
+///
+/// ```
+/// use bevy_ecs::{prelude::*, component::WriteAccess};
+///
+/// // A set of components that we want to restrict mutable access to.
+/// // By marking the components with this type, they can only be mutated
+/// // in places that `Marker` can be named.
+/// struct Marker;
+///
+/// #[derive(Component)]
+/// #[component(vis = "Marker")]
+/// pub struct Chassis;
+///
+/// #[derive(Component)]
+/// #[component(vis = "Marker")]
+/// pub struct Axle {
+///     // The fields can be public, since no one outside this module can get mutable access
+///     // once this component has been inserted into the world.
+///     pub torque: f64,
+/// }
+///
+/// #[derive(Component)]
+/// #[component(vis = "Marker")]
+/// pub struct Tires {
+///     pub width: f64,
+/// }
+/// #
+/// # use bevy_ecs::component::ProtectedBundle;
+///
+/// // In order to include the components in a bundle, we must encapsulate the `Marker` type.
+///
+/// #[derive(Bundle)]
+/// pub struct Automobile {
+///     // Flatten the three smaller bundles to make one big bundle.
+///     #[bundle]
+///     chassis: ProtectedBundle<Chassis, Marker>,
+///     #[bundle]
+///     axle: ProtectedBundle<Axle, Marker>,
+///     #[bundle]
+///     tires: ProtectedBundle<Tires, Marker>,
+/// }
+/// #
+/// // `new` method omitted.
+/// # impl Automobile {
+/// #     pub fn new(axle: Axle, tires: Tires) -> Self {
+/// #         Self {
+/// #             chassis: ProtectedBundle::new(Chassis),
+/// #             axle: ProtectedBundle::new(axle),
+/// #             tires: ProtectedBundle::new(tires),
+/// #         }
+/// #     }
+/// # }
+///
+/// fn setup(mut commands: Commands) {
+///     commands.spawn_bundle(
+///         Automobile::new(Axle { torque: 100. }, Tires { width: 20. })
+///     );
+/// }
+/// #
+/// # bevy_ecs::system::assert_is_system(setup);
+/// ```
+pub struct ProtectedBundle<T, Marker: 'static>
+where
+    T: WriteComponent<Marker>,
+{
+    // FIXME: Derive this when we can do #[bundle(ignore)].
```

Comment:
Would be a good idea to like the issue on comment? Don't know if this is a good good practice tho